flang/lib/Lower/IO.cpp

```diff
@@ -141,162 +151,258 @@ static mlir::Value genEndIO(Fortran::lower::AbstractConverter &converter,
 static void makeNextConditionalOn(Fortran::lower::FirOpBuilder &builder,
-                                  mlir::Location loc, mlir::Value ok,
-                                  mlir::OpBuilder::InsertPoint &insertPt) {
-  if (!ok)
+                                  mlir::Location loc,
+                                  mlir::OpBuilder::InsertPoint &insertPt,
+                                  bool checkResult, mlir::Value ok,
+                                  bool inIterWhileLoop = false) {
+  if (!checkResult || !ok)
+    // Either I/O calls do not need to be checked, or the next I/O call is the
+    // first potentially fallable call.
     return;
-  // Last call returned the bool `ok`.
-  auto where = builder.create<fir::WhereOp>(loc, ok, /*withOtherwise=*/false);
+  // A previous I/O call for a statement returned the bool `ok`.  If this call
+  // is in a fir.iterate_while loop, the result must be propagated up to the
+  // loop scope.  That is done in genIoLoop, but it is enabled here.
+  auto whereOp =
+      inIterWhileLoop
+          ? builder.create<fir::WhereOp>(loc, builder.getI1Type(), ok, true)
+          : builder.create<fir::WhereOp>(loc, ok, /*withOtherwise=*/false);
   if (!insertPt.isSet())
     insertPt = builder.saveInsertionPoint();
-  builder.setInsertionPointToStart(&where.whereRegion().front());
-}
-
-/// Get the OutputXyz routine to output a value with type `type`.
-static mlir::FuncOp getOutputRuntimeFunc(Fortran::lower::FirOpBuilder &builder,
-                                         mlir::Type type) {
-  if (auto ty = type.dyn_cast<mlir::IntegerType>()) {
-    if (ty.getWidth() == 1)
-      return getIORuntimeFunc<mkIOKey(OutputLogical)>(builder);
-    return getIORuntimeFunc<mkIOKey(OutputInteger64)>(builder);
-  }
-  if (auto ty = type.dyn_cast<mlir::FloatType>()) {
-    if (ty.getWidth() <= 32)
-      return getIORuntimeFunc<mkIOKey(OutputReal32)>(builder);
-    return getIORuntimeFunc<mkIOKey(OutputReal64)>(builder);
-  }
-  if (auto ty = type.dyn_cast<fir::CplxType>()) {
-    if (ty.getFKind() <= 4)
-      return getIORuntimeFunc<mkIOKey(OutputComplex32)>(builder);
-    return getIORuntimeFunc<mkIOKey(OutputComplex64)>(builder);
-  }
+  builder.setInsertionPointToStart(&whereOp.whereRegion().front());
+}
+
+template <typename D>
+static void genIoLoop(Fortran::lower::AbstractConverter &converter,
+                      mlir::Value cookie, const D &ioImpliedDo,
+                      bool checkResult, mlir::Value &ok, bool inIterWhileLoop);
+
+/// Get the OutputXyz routine to output a value of the given type.
+static mlir::FuncOp getOutputFunc(Fortran::lower::FirOpBuilder &builder,
+                                  mlir::Type type) {
+  if (auto ty = type.dyn_cast<mlir::IntegerType>())
+    return ty.getWidth() == 1
+               ? getIORuntimeFunc<mkIOKey(OutputLogical)>(builder)
+               : getIORuntimeFunc<mkIOKey(OutputInteger64)>(builder);
```

Comment:
Does LLVM prefer using ternary ops?

Reply:
I don't think llvm comments one way or the other.  I prefer the ternary ops here because they nicely highlight the two-level selection choices, and in my opinion, make the code easier to read.  There is also a small bonus of slightly fewer lines of code, which also tends to improve understanding.

Reply:
Ok. Some people really have a strong dislike of it.  The pattern
```c++
if (...)
  return ...;
return ...;
```
is used ubiquitously in LLVM. So when in Rome ...